### PR TITLE
[Knet driver] Upgrade Broadcom xgs saibcm-modules to 10.1

### PR DESCRIPTION
**Pull request by Gfrom2016**
#### Why I did it
Upgrade xgs saibcm-modules to 10.1.

##### Work item tracking
- Microsoft ADO **(number only)**: 28240459

#### How I did it
Generate diff patch from SAI SDK 10.1 repo and apply to platform/broadcom/saibcm-modules/.

#### How to verify it
Build a master image and run SONiC test with it.
TH2 7260cx3:  [Test plan 66b460e40fde6eb55e3ad994: TfsGit_Manual_PR_None_BUILD_614424_JOB_7260cx3.t1.202311.NightlyTest_by_Elastictest - Elastictest (elastictest.org)](https://elastictest.org/scheduler/testplan/66b460e40fde6eb55e3ad994)
TD3 7050cx3: [Pipelines - Run SAI_RELEASE_PRE_TEST_PRE_TEST_FOR_SAI_RELEASE_internal_591881_20240710.1 (azure.com)](https://dev.azure.com/mssonic/internal/_build/results?buildId=591881&view=results)
TH 7060cx: [Test plan 66b999b8e664a2241593dccf: upgrade_saibcm_10.1_7060cx.t1.202311.NightlyTest_by_Elastictest - Elastictest (elastictest.org)](https://elastictest.org/scheduler/testplan/66b999b8e664a2241593dccf)

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305
- [x] 202311
- [x] 202405

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by Gfrom2016** [comment]
@adyeung Could you help review this PR? Thanks!

**Comment by adyeung** [comment]
Adding BRCM reviewer @michaelli10

**Comment by Gfrom2016** [comment]
@michaelli10 @chravikumar @rajanidadi Could you help review this PR? Thanks.

File: platform/broadcom/saibcm-modules/sdklt/linux/bcmgenl/bcmgenl_psample.c [hunk 1]
```diff
@@ -0,0 +1,1076 @@
...
+                (last_skb, "%s: failed to alloc generic mem for pkt skb: %lu\n",
+                 __func__, g_bcmgenl_psample_stats.pkts_d_no_mem);
+            goto PSAMPLE_FILTER_CB_PKT_HANDLED;
+        }
+
+        /* SDKLT-43751: Use ptr offset to pkt payload to send to genetlink */
+        /* setup skb to point to pkt */
+        memcpy(skb_psample->data, pkt_ptr, meta.trunc_size);
+        skb_put(skb_psample, meta.trunc_size);
+        skb_psample->len = meta.trunc_size;
```
[the remark this answers is not in the document]
Comment (Gfrom2016) [reply platform/broadcom/saibcm-modules/sdklt/linux/bcmgenl/bcmgenl_psample.c]: fixed as you suggested. Can you also tell me which version of SAI has this fix?

File: platform/broadcom/saibcm-modules/systems/linux/kernel/modules/bcm-genl/bcm-genl-psample.c [hunk 2]
```diff
@@ -160,239 +144,155 @@ typedef struct psample_work_s {
...
-
-    /* only sample rx reason set, consume pkt */
-    return (1);
+    return 0;
 }
 
 static int
-psample_meta_get(int unit, uint8_t *pkt, void *pkt_meta, psample_meta_t *sflow_meta)
+psample_meta_get(int dev_no, kcom_filter_t *kf, void *pkt_meta,
+                 psample_meta_t *sflow_meta)
```
[the remark this answers is not in the document]
Comment (Gfrom2016) [reply platform/broadcom/saibcm-modules/systems/linux/kernel/modules/bcm-genl/bcm-genl-psample.c:232]: fixed as you suggested.

**Review by michaelli10 — APPROVED** [review]

**Pull request merged**